File: app-android/vela-wallet/app/src/main/java/app/getvela/wallet/feature/browser/core/RequestRouter.kt
package app.getvela.wallet.feature.browser.core

import org.json.JSONArray
import org.json.JSONObject

/**
 * Answers a forwarded request the shell routes itself (spec 044 T023; the
 * desktop's `page.rs` `Route::State / Switch / Ack / Read / Unsupported`
 * arms). A signature goes to the signing controller (phase 4); everything
 * else is answered here — a wallet fact from the browser's chain, a read
 * through the person's own pool, an acknowledgement, or a refusal.
 */
class RequestRouter(
    private val ports: Ports,
) {
    interface Ports {
        /** The chain the browser is on. */
        fun browserChain(): Int

        /** The chains this wallet has (the settings machine's rows). */
        fun knownChains(): List<Int>

        /** The browser moves to a known chain: remembered, and the permissions machine told. */
        fun switchChain(chainId: Int)

        /** `eth_*` through the pool: the JSON-RPC body (`result` or `error`), or `null` when nothing answered. */
        suspend fun poolCall(chainId: Int, method: String, params: JSONArray, bundler: Boolean): JSONObject?

        fun respond(id: String, json: JSONObject)

        /** A signature request: the signing machines' (phase 4). */
        fun sign(id: String, method: String, paramsJson: String, origin: String)

        /**
         * A user-operation hash this wallet submitted for a page: the page
         * was answered with it as its "transaction hash" (the web's
         * non-blocking rule), so its receipt lookups are translated —
         * `null` when the hash is not one of ours.
         */
        suspend fun receiptFor(userOpHash: String): Receipt?
    }

    sealed class Receipt {
        data object Pending : Receipt()
        data class Landed(val txHash: String) : Receipt()
    }

    suspend fun route(id: String, method: String, paramsJson: String, origin: String) {
        when (val route = DappRpc.route(method)) {
            DappRpc.Route.Sign -> ports.sign(id, method, paramsJson, origin)
            // The chain the wallet is on, in the two notations the two methods
            // are specified in. `net_version` is decimal — a hex answer there
            // is a string comparison every dApp fails.
            DappRpc.Route.State -> ports.respond(
                id,
                BrowserExecutor.resultJson(id, if (method == "net_version") ports.browserChain().toString() else DappRpc.hexChainId(ports.browserChain())),
            )
            DappRpc.Route.Switch -> {
                val wanted = DappRpc.switchChainParam(paramsJson)
                when {
                    wanted == null -> ports.respond(id, BrowserExecutor.errorJson(id, -32602, "Invalid chainId"))
                    wanted !in ports.knownChains() -> ports.respond(id, BrowserExecutor.errorJson(id, 4902, "Unrecognized chain ID"))
                    else -> {
                        ports.switchChain(wanted)
                        ports.respond(id, BrowserExecutor.resultJson(id, null))
                    }
                }
            }
            // Acknowledged, and nothing changed. A network or a token is added
            // in this wallet's own settings, by a person looking at it — never
            // because a page asked while it had the floor.
            DappRpc.Route.Ack -> ports.respond(id, BrowserExecutor.resultJson(id, null))
            is DappRpc.Route.Read -> {
                var params = runCatching { JSONArray(paramsJson) }.getOrElse { JSONArray() }
                // A receipt asked for by a user-operation hash we answered
                // with: pending → nothing yet; landed → the node's receipt
                // for the real transaction (spec 028's translation).
                if (method == "eth_getTransactionReceipt" || method == "eth_getTransactionByHash") {
                    val asked = params.optString(0)
                    when (val receipt = asked.takeIf { it.startsWith("0x") }?.let { ports.receiptFor(it) }) {
                        RequestRouter.Receipt.Pending -> { ports.respond(id, BrowserExecutor.resultJson(id, null)); return }
                        is RequestRouter.Receipt.Landed -> params = JSONArray().put(receipt.txHash)
                        null -> Unit
                    }
                }
                val body = ports.poolCall(ports.browserChain(), method, params, route.bundler)
                val error = body?.optJSONObject("error")
                when {
                    body == null -> ports.respond(id, BrowserExecutor.errorJson(id, -32603, "No endpoint answered"))
                    error != null -> ports.respond(id, BrowserExecutor.errorJson(id, error.optInt("code", -32603), error.optString("message").ifBlank { "The node refused this call" }))
                    else -> ports.respond(id, BrowserExecutor.resultJson(id, body.opt("result")))
                }
            }
            // 4900 and not 4001: the person did not decline, and a dApp that
            // reads a decline where there was none tells them they refused
            // something they never saw.
            DappRpc.Route.Unsupported -> ports.respond(id, BrowserExecutor.errorJson(id, 4900, "Vela cannot answer $method yet"))
        }
    }
}

File: app-ios/VelaWallet/VelaWallet/Features/Explore/Core/RequestRouter.swift
//
//  RequestRouter.swift
//  VelaWallet
//
//  What happens to a request the permissions machine forwarded but does not
//  answer itself: the reads, the state questions, the chain switch, the
//  acknowledgements, and the refusals.
//
//  Ported from `app-android/.../feature/browser/core/RequestRouter.kt`
//  (spec 044), which is the desktop's `executor/dapp_rpc.rs`.
//
//  ## `net_version` is decimal and `eth_chainId` is hex
//
//  Not a nicety. A hex answer to `net_version` is a string comparison every
//  dApp fails, and the failure is silent: the site simply believes it is on a
//  different network than the wallet.
//
//  ## A user-op hash is not a transaction hash, and a page must not learn that
//
//  `eth_sendTransaction` is answered with a **transaction** hash. A page that
//  then polls `eth_getTransactionReceipt` for a hash this wallet minted as a
//  user operation gets the receipt for the transaction that carried it — the
//  translation spec 028 found was needed when a real dApp's swap "submitted
//  but never confirmed".
//

import Foundation

@MainActor
final class RequestRouter {

    enum Receipt {
        case pending
        case landed(txHash: String)
    }

    struct Ports {
        var browserChain: () -> Int = { 100 }
        var knownChains: () -> [Int] = { [] }
        var switchChain: (Int) -> Void = { _ in }
        /// One call through the wallet's own pool. `nil` = no endpoint
        /// answered at all, which is different from a node that refused.
        var poolCall: (_ chainId: Int, _ method: String, _ params: [Any], _ bundler: Bool) async -> [String: Any]?
        = { _, _, _, _ in nil }
        var respond: (_ id: String, _ json: [String: Any]) -> Void = { _, _ in }
        var sign: (_ id: String, _ method: String, _ paramsJson: String, _ origin: String) -> Void
        = { _, _, _, _ in }
        /// `nil` when this hash is not one of ours.
        var receiptFor: (_ userOpHash: String) async -> Receipt? = { _ in nil }
    }

    var ports: Ports

    init(ports: Ports = Ports()) {
        self.ports = ports
    }

    func route(id: String, method: String, paramsJson: String, origin: String) async {
        let route = DappRpc.route(method)

        switch route {
        case .sign:
            ports.sign(id, method, paramsJson, origin)

        case .state:
            let chainId = ports.browserChain()
            let answer: Any = method == "net_version"
                ? String(chainId)
                : DappRpc.hexChainId(chainId)
            ports.respond(id, BrowserExecutor.resultJson(id: id, result: answer))

        case .switchChain:
            guard let wanted = DappRpc.switchChainParam(paramsJson) else {
                ports.respond(id, BrowserExecutor.errorJson(id: id, code: -32602, message: "Invalid chainId"))
                return
            }
            guard ports.knownChains().contains(wanted) else {
                // 4902 is the standard's "add it first", and adding a network
                // by a page's request is 055's path, not this one's.
                ports.respond(id, BrowserExecutor.errorJson(id: id, code: 4902, message: "Unrecognized chain ID"))
                return
            }
            ports.switchChain(wanted)
            ports.respond(id, BrowserExecutor.resultJson(id: id, result: nil))

        case .ack:
            // Acknowledged, and nothing changed. Saying "no" to
            // `wallet_addEthereumChain` breaks dApps that call it defensively
            // before every request; acting on it would let a page add a
            // network nobody vetted.
            ports.respond(id, BrowserExecutor.resultJson(id: id, result: nil))

        case .read(let bundler):
            await read(id: id, method: method, paramsJson: paramsJson, bundler: bundler)

        case .unsupported:
            // **4900, not 4001.** The person did not decline.
            ports.respond(id, BrowserExecutor.errorJson(
                id: id, code: 4900, message: "Vela cannot answer \(method) yet"
            ))
        }
    }

    private func read(id: String, method: String, paramsJson: String, bundler: Bool) async {
        var params = Self.array(paramsJson)

        if method == "eth_getTransactionReceipt" || method == "eth_getTransactionByHash",
           let asked = params.first as? String, asked.hasPrefix("0x") {
            switch await ports.receiptFor(asked) {
            case .pending:
                // Ours, and not landed. `null` is what a node says about a
                // transaction it has not seen, which is exactly true here.
                ports.respond(id, BrowserExecutor.resultJson(id: id, result: nil))
                return
            case .landed(let txHash):
                params = [txHash]
            case nil:
                break
            }
        }

        let body = await ports.poolCall(ports.browserChain(), method, params, bundler)
        guard let body else {
            ports.respond(id, BrowserExecutor.errorJson(
                id: id, code: -32603, message: "No endpoint answered"
            ))
            return
        }
        if let error = body["error"] as? [String: Any] {
            let message = (error["message"] as? String).flatMap { $0.isEmpty ? nil : $0 }
                ?? "The node refused this call"
            ports.respond(id, BrowserExecutor.errorJson(
                id: id,
                code: (error["code"] as? NSNumber)?.intValue ?? -32603,
                message: message
            ))
            return
        }
        ports.respond(id, BrowserExecutor.resultJson(id: id, result: body["result"]))
    }

    static func array(_ json: String) -> [Any] {
        guard let data = json.data(using: .utf8),
              let params = try? JSONSerialization.jsonObject(with: data) as? [Any]
        else { return [] }
        return params
    }
}
